sql/moz-fx-data-shared-prod/org_mozilla_fenix_derived/experiment_search_events_live_v1/init.sql

```diff
@@ -8,24 +8,91 @@ IF
   OPTIONS
     (enable_refresh = TRUE, refresh_interval_minutes = 5)
   AS
-  WITH fenix AS (
+  -- Materialized views do not support sub-queries for UNNESTing fields.
+  -- This limitation prevents summing up individual nested counters.
+  -- As a workaround the nested counters are converted into arrays of counts. The
+  -- conversion is done by a combination of dumping the fields as JSON
+  -- strings, applying regexes and concatenating extracted strings.
+  --
+  -- (1) Dump the nested counter field as JSON string, for example:
+  --     [{"key":"duckduckgo","value":7}, {"key":"google","value":4}]
+  -- (2) Using a regex, extract all values and dump the result array as
+  --     JSON string: ["7", "4"]
+  -- (3) Using regex replace, transform the count numbers in the array
+  --     to an array where the position of the element determines the
+  --     count type [search_ad_clicks, search_with_ads, search_count]:
+  --     "[[7, 0, 0], [4, 0, 0]]"
+  -- (4) Convert the nested array string to an array and concatenate
+  --     nested arrays of ping to a single array and dump as JSON string,
+  --     for example: "[[7, 0, 0], [4, 0, 0], [0, 1, 0], [0, 0, 3]]"
+  --
+  -- (5) Convert the nested array string into an array, unnest and sum
+  --     individual counts based on their array index:
+  --        ad_clicks_count = 11
+  --        search_with_ads_count = 1
+  --        search_count = 3
```

Comment:
This is a hairy workaround. I'm hoping this step by step way of explaining what is happening here does help. I'm also open to suggestions for better solutions. At least this does produce the correct results.

Reply:
What if we just had separate materialized views for each metric type? That shouldn't lead to inflated counts, and would be more readable. Downside is some repeated code.

Reply:
That would also be an option, though we'd end up having 12 materialized views then ([`org_mozilla_fenix`, `telemetry`, `org_mozilla_firefox_beta`, `org_mozilla_firefox_beta`] * [`ad_clicks_count`, `search_with_ads_count`, `search_count`])

Reply:
Could we deploy those within a `sql.py` file?

Reply:
I really appreciate the thought that's gone into the documentation for this workaround. In particular, marking lines with `(3)`, etc. to refer to the logical steps makes parsing it much easier.

That said, it still would be pretty difficult to modify or debug this code if we find the results don't look right.

I think we have sufficient motivation at this point for generating these queries, which could make maintaining 12 views feasible. As long as we make sure all the generated files contain a comment refer referring to the script that created them, it should be reasonably comprehensible.

Reply:
I ended up implementing a script for generating these queries. One limitation I didn't consider and only noticed when I test deployed the views is: https://cloud.google.com/bigquery/docs/materialized-views-faq#join_and_union_all

When we have multiple materialized views that query from the same base table (as we do for every metric) and then `JOIN` or `UNION ALL` these views, the actual base tables get queried from. Which is expensive. I added a comment here for more context: https://github.com/mozilla/bigquery-etl/pull/1904#discussion_r597216365

This issue doesn't occur in the implementation here, which now seems to be preferable?

Reply:
This is fascinating.

> This issue doesn't occur in the implementation here, which now seems to be preferable?

I agree with this conclusion. It sounds like we need to have only one materialized view per table, and string manipulation is the only viable option.

Reply:
Can the endpoint issue separate queries? I'd imagine this is going into Grafana, in which case these would end up in separate plots?

Reply:
Besides the Grafana plots, the data is and will also be used in additional monitoring queries to show warnings to users. Currently, we only have enrollment related queries, for example, to show a warning if the number of enrollments is too low or the number of unenrollments is too high. There'll be more such queries in the future (a proposal is in the works for Q2) which also might look at search metrics and possibly have queries that end up querying from different materialized views with the same base table again.

Reply:
I see - I'll let you speak to whether it's doable to split those. Given it's not, this code is clearly the path forward!